### PR TITLE
[dbus] add `retain_active_session` option for deactivate API

### PR DESCRIPTION
**Pull request by mia1yang**

### CODE REVIEW

**Comment by bukepo** [comment]
Should disconnect be the default behavior? Perhaps adding an option "retain-active-session"?

**Comment by wgtdkp** [comment]
> Should disconnect be the default behavior? Perhaps adding an option "retain-active-session"?

+1

**Comment by mia1yang** [comment]
> Should disconnect be the default behavior? Perhaps adding an option "retain-active-session"?

Change `force disconnect` to be the default logic now. Thanks~

File: src/dbus/server/introspect.xml [hunk 1]
```diff
@@ -246,8 +246,14 @@
       <arg name="epskc" type="s" direction="out"/>
     </method>
 
-    <!-- DeactivateEphemeralKeyMode: Deactivate ePSKc mode. -->
+    <!-- DeactivateEphemeralKeyMode: Deactivate ePSKc mode.
+      @retain_active_session: 
+        "false" - Disconnects the Border Agent from any active secure sessions.
+                  Ephemeral key would be cleared after the session is disconnected.
+        "true" - Clears ephemeral key only when Border Agent state is "Start".
```
Comment (librasungirl) [inline src/dbus/server/introspect.xml:253]:
```suggestion
        "true" - Clears ephemeral key only when Border Agent state is running and no active session established
        
        Note that when skipped, it's equivalent with "false" condition, active session will be closed if any. 
```

**Review by librasungirl — APPROVED** [review]

**Comment by bukepo** [comment]
missing test?

**Comment by mia1yang** [comment]
> missing test?

Test is in openthread branch, and it seems no depends-on support on otbr (only ot have depends-on of otbr).
Need to merge openthread change firstly to clear the CI failure.

File: src/dbus/server/dbus_thread_object_rcp.cpp [hunk 2]
```diff
@@ -2009,22 +2009,33 @@ void DBusThreadObjectRcp::DeactivateEphemeralKeyModeHandler(DBusRequest &aReques
...
+                break;
+            case OT_BORDER_AGENT_STATE_STARTED:
+                otBorderAgentClearEphemeralKey(threadHelper->GetInstance());
+                break;
+            }
+            ExitNow();
+        }
     }
 
+    otBorderAgentDisconnect(threadHelper->GetInstance());
```
Comment (sunytt) [inline src/dbus/server/dbus_thread_object_rcp.cpp:2037]: I think `otBorderAgentClearEphemeralKey` also needs to be called here, if there is no secure session, calling `otBorderAgentDisconnect` wouldn't clear the ephemeral key.

File: src/dbus/server/introspect.xml (hunk 1 above)
Comment (mia1yang) [reply src/dbus/server/introspect.xml]: Thanks! Done.

File: src/dbus/server/dbus_thread_object_rcp.cpp (hunk 2 above)
Comment (mia1yang) [reply src/dbus/server/dbus_thread_object_rcp.cpp]: Thanks! Modified according to BA state now.

File: src/dbus/server/introspect.xml [hunk 3]
```diff
@@ -246,8 +246,15 @@
       <arg name="epskc" type="s" direction="out"/>
     </method>
 
-    <!-- DeactivateEphemeralKeyMode: Deactivate ePSKc mode. -->
+    <!-- DeactivateEphemeralKeyMode: Deactivate ePSKc mode.
+      @retain_active_session:
```
Comment (sunytt) [inline src/dbus/server/introspect.xml:250]: could mention the default value is false?

**Review by sunytt — APPROVED** [review]

File: src/dbus/server/dbus_thread_object_rcp.cpp [hunk 4]
```diff
@@ -2018,7 +2020,14 @@ void DBusThreadObjectRcp::DeactivateEphemeralKeyModeHandler(DBusRequest &aReques
         error = OT_ERROR_FAILED;
```
Comment (wgtdkp) [inline src/dbus/server/dbus_thread_object_rcp.cpp:2020]: Why it is an error that border agent is stopped when trying to deactivate the ePSKc mode? Can't ePSKc be deactivated when border agent is stopped? I think this `DBusThreadObjectRcp::DeactivateEphemeralKeyModeHandler` should be implemented without checking the border agent state. For example:

```
void DBusThreadObjectRcp::DeactivateEphemeralKeyModeHandler(DBusRequest &aRequest)
{
    otError error        = OT_ERROR_NONE;
    auto    threadHelper = mHost.GetThreadHelper();
    bool    retain_active_session;
    auto    args = std::tie(retain_active_session);

    VerifyOrExit(mBorderAgent.GetEphemeralKeyEnabled(), error = OT_ERROR_NOT_CAPABLE);

    if ((DBusMessageToTuple(*aRequest.GetMessage(), args) == OTBR_ERROR_NONE) && !retain_active_session)
    {
        otBorderAgentDisconnect(threadHelper->GetInstance());
    }
    otBorderAgentClearEphemeralKey(threadHelper->GetInstance());

exit:
    aRequest.ReplyOtResult(error);
}
```

File: src/dbus/server/introspect.xml (hunk 1 above)
Comment (librasungirl) [reply src/dbus/server/introspect.xml]: "Note" part is missed?

File: src/dbus/server/dbus_thread_object_rcp.cpp (hunk 4 above)
Comment (mia1yang) [reply src/dbus/server/dbus_thread_object_rcp.cpp]: Accepted. The key diffs are removing `OT_ERROR_FAILED` and `OT_ERROR_INVALID_STATE` to keep deactivate ePSKc successful. Add log for production debug.

File: src/dbus/server/introspect.xml (hunk 3 above)
Comment (mia1yang) [reply src/dbus/server/introspect.xml:250]: Thanks，accepted.

File: src/dbus/server/introspect.xml [hunk 5]
```diff
@@ -246,8 +246,15 @@
       <arg name="epskc" type="s" direction="out"/>
     </method>
 
-    <!-- DeactivateEphemeralKeyMode: Deactivate ePSKc mode. -->
+    <!-- DeactivateEphemeralKeyMode: Deactivate ePSKc mode.
+      @retain_active_session: 
+        "false" - Default value. Disconnects the Border Agent from any active secure sessions.
+                  Ephemeral key would be cleared after the session is disconnected.
+        "true" - Clears ephemeral key only when Border Agent state is running
+                 and no active session established.
```
Comment (wgtdkp) [inline src/dbus/server/introspect.xml:254]: The description is not accurate. The ephemeral key is always cleared no matter there is active session or not.

File: src/dbus/server/introspect.xml [hunk 6]
```diff
@@ -246,8 +246,15 @@
       <arg name="epskc" type="s" direction="out"/>
     </method>
 
-    <!-- DeactivateEphemeralKeyMode: Deactivate ePSKc mode. -->
+    <!-- DeactivateEphemeralKeyMode: Deactivate ePSKc mode.
+      @retain_active_session: 
+        "false" - Default value. Disconnects the Border Agent from any active secure sessions.
```
Comment (wgtdkp) [inline src/dbus/server/introspect.xml:251]: Why "Default value" here? I think line 2017 in src/dbus/server/dbus_thread_object_rcp.cpp is making the "retain_active_session" a mandatory argument. Given it is not an optional value, there is no value to mention about the default value because the client always need to provide the argument explicitly.

Comment (mia1yang) [reply src/dbus/server/introspect.xml]: Thanks, changed the description to map the latest code.

File: src/dbus/server/introspect.xml (hunk 5 above)
Comment (mia1yang) [reply src/dbus/server/introspect.xml]: Modified accordingly. Thanks!

File: src/dbus/server/introspect.xml [hunk 7]
```diff
@@ -246,8 +246,16 @@
       <arg name="epskc" type="s" direction="out"/>
     </method>
 
-    <!-- DeactivateEphemeralKeyMode: Deactivate ePSKc mode. -->
+    <!-- DeactivateEphemeralKeyMode: Deactivate ePSKc mode.
+      @retain_active_session: 
+        "false" - Disconnects the Border Agent from any active secure sessions.
+                  Ephemeral key would be cleared after the session is disconnected.
+        "true" - Deactivate ephemeral key mode only when Border Agent state is running
+                 and no active session established. If a commissioner connection is active,
```
Comment (wgtdkp) [inline src/dbus/server/introspect.xml:254]: > Deactivate ephemeral key mode only when Border Agent state is running
                 and no active session established

This is not accurate, the ephemeral key is always deactivated even if the commissioner is connected.

**Review by wgtdkp — APPROVED** [review]

File: src/dbus/server/introspect.xml (hunk 7 above)
Comment (mia1yang) [reply src/dbus/server/introspect.xml]: Modified accordingly, reopen if not proper. Thanks~

File: src/dbus/server/introspect.xml [hunk 8]
```diff
@@ -246,8 +246,15 @@
       <arg name="epskc" type="s" direction="out"/>
     </method>
 
-    <!-- DeactivateEphemeralKeyMode: Deactivate ePSKc mode. -->
+    <!-- DeactivateEphemeralKeyMode: Deactivate ePSKc mode.
+      @retain_active_session: 
+        "false" - Disconnects the Border Agent from any active secure sessions.
+                  Ephemeral key would be cleared after the session is disconnected.
+        "true" - Deactivate ephemeral key mode softly. If a commissioner connection is active,
+                 wait for it to disconnect or for the ephemeral key timeout.
```
Comment (wgtdkp) [inline src/dbus/server/introspect.xml:254]: Sorry, I think the second sentence is also not accurate. I would rephrase as:

```
"true" - Deactivate ephemeral key mode softly. If there is already an active commissioner connection, the session will be retained
```

Comment (mia1yang) [reply src/dbus/server/introspect.xml]: Changed accordingly to reduce misleading, thanks!

**Comment by mia1yang** [comment]
The CI failure of this PR is due to Dbus API new option added, but the original unit test in OT has not added the new option. The CI failure should be clear after PR `https://github.com/openthread/openthread/pull/10829` merged. Highlight error info here: "ERROR:dbus.connection:Unable to set arguments () according to signature 'b':"

File: src/dbus/server/dbus_thread_object_rcp.cpp [hunk 9]
```diff
@@ -2009,21 +2009,17 @@ void DBusThreadObjectRcp::DeactivateEphemeralKeyModeHandler(DBusRequest &aReques
...
-        break;
-    case OT_BORDER_AGENT_STATE_ACTIVE:
-        error = OT_ERROR_INVALID_STATE;
-        break;
-    case OT_BORDER_AGENT_STATE_STARTED:
-        otBorderAgentClearEphemeralKey(threadHelper->GetInstance());
-        break;
+        otBorderAgentDisconnect(threadHelper->GetInstance());
     }
+    otBorderAgentClearEphemeralKey(threadHelper->GetInstance());
```
Comment (librasungirl) [inline src/dbus/server/dbus_thread_object_rcp.cpp:2022]: To better match the doc in xml, may update as below?
 ```
if (!retain_active_session && otBorderAgentGetState(threadHelper->GetInstance()) >= OT_BORDER_AGENT_STATE_ACTIVE)
 {
    // Ephemeral Key will be cleared after the session is disconnected
     otBorderAgentDisconnect(threadHelper->GetInstance());
 }
else
{
    otBorderAgentClearEphemeralKey(threadHelper->GetInstance());
}
```

Comment (mia1yang) [reply src/dbus/server/dbus_thread_object_rcp.cpp:2022]: otBorderAgentClearEphemeralKey also needs to be called after disconnect(when BA is start state), if there is no secure session, calling otBorderAgentDisconnect wouldn't clear the ephemeral key.

Comment (librasungirl) [reply src/dbus/server/dbus_thread_object_rcp.cpp:2022]: it's covered in the` else` already - when no secure session, otBorderAgentClearEphemeralKey will be called

Comment (sunytt) [reply src/dbus/server/dbus_thread_object_rcp.cpp:2022]: I think it's ok to let otBorderAgentDisconnect take care of the different BA states (e.g. no-op when BA state is not `OT_BORDER_AGENT_STATE_ACTIVE`), the current code looks good and simple to me.

**Review by sunytt — APPROVED** [review]
LGTM, thanks

File: src/dbus/server/dbus_thread_object_rcp.cpp (hunk 9 above)
Comment (librasungirl) [reply src/dbus/server/dbus_thread_object_rcp.cpp:2022]: As offline discussion, fine to keep it as is, and may update API reference for otBorderAgentClearEphemeralKey and otBorderAgentDisconnect for forcely deactivation operation

**Review by librasungirl — APPROVED** [review]

**Review by Irving-cl — APPROVED** [review]
LGTM 👍 

One small suggestion: it would be great to write PR description. Probably with a one-sentence summary and some explanation about the background,

**Review by jwhui — APPROVED** [review]
